### PR TITLE
discovery: don't send private chan updates to wider network

### LINKED ISSUE #2904: The updatechanpolicy propogates fee rates of private channels to other nodes in ChannelUpdate messages
### Background

When i do `updatechanpolicy` command to change fee of **private channel** this channel fees are propogated to other nodes in ChannelUpdate messages.

### Your environment

* version of `lnd` - 0.5.2-99-beta commit=queue/v1.0.1-269-gfa73a60167aa07df948020eb5f1f31893a113f9c
* which operating system (`uname -a` on *Nix) CentOS 6
* version of `btcd`, `bitcoind`, or other backend  - 0.17.1

### Actual behaviour

I have BLW wallet, it has channel `02db9eb1e37acbdc9a987eaf878bd3c3d5f5c9afa2159310e048d816f344d29358` (`570351:887:1`) with `LND-12`
In `LND-12` i have run trace debug level to see ChannelUpdate messages.

I did:

    l updatechanpolicy 0 0.000500 30 489d625be31c41fbac4c51a3a5b73a2540b0babfa140b0b7c0d406f3cf19e114:1

But i found that ChannelUpdate messages propogates the fee of this private channel to other nodes. But this channel is private. I think it's bug.

This channel as:

    {
        "active": false,
        "remote_pubkey": "02db9eb1e37acbdc9a987eaf878bd3c3d5f5c9afa2159310e048d816f344d29358",
        "channel_point": "489d625be31c41fbac4c51a3a5b73a2540b0babfa140b0b7c0d406f3cf19e114:1",
        "chan_id": "627107556471799809",
        "capacity": "4000000",
        "local_balance": "3980822",
        "remote_balance": "0",
        "commit_fee": "19178",
        "commit_weight": "600",
        "fee_per_kw": "26489",
        "unsettled_balance": "0",
        "total_satoshis_sent": "0",
        "total_satoshis_received": "0",
        "num_updates": "0",
        "pending_htlcs": [
        ],
        "csv_delay": 2016,
        "private": true,
        "initiator": true,
        "chan_status_flags": "ChanStatusDefault"
    },

ChannelUpdate about fee of BLW to channel BLW &lt;--&gt; LND-12 (IP addresses was changed here):

    2019-04-06 20:31:19.412 [TRC] PEER: writeMessage to 02db9eb1e37acbdc9a987eaf878bd3c3d5f5c9afa2159310e048d816f344d29358@XX.XX.XX.XX:46183: (*lnwire.ChannelUpdate)(0xc05f2ba2c0)({
     Signature: (lnwire.Sig) (len=64 cap=64) {
      00000000  5d 52 dc 62 bf bc 8b 5e  bb 84 db cf a4 b6 89 0a  |]R.b...^........|
      00000010  04 60 5b 12 61 fb fb 3c  41 4a 53 b3 18 4b 9b fb  |.`[.a..<AJS..K..|
      00000020  0f c0 a2 77 7d 26 f7 74  9c 4b d3 75 4a 8c 1b 21  |...w}&.t.K.uJ..!|
      00000030  81 91 43 f3 3f 1b d2 5d  5b 09 0b f8 7e db d3 c3  |..C.?..][...~...|
     },
     ChainHash: (chainhash.Hash) (len=32 cap=32) 000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f,
     ShortChannelID: (lnwire.ShortChannelID) 570351:887:1,
     Timestamp: (uint32) 1554575479,
     MessageFlags: (lnwire.ChanUpdateMsgFlags) 00000001,
     ChannelFlags: (lnwire.ChanUpdateChanFlags) 00000001,
     TimeLockDelta: (uint16) 30,
     HtlcMinimumMsat: (lnwire.MilliSatoshi) 1000 mSAT,
     BaseFee: (uint32) 0,
     FeeRate: (uint32) 500,
     HtlcMaximumMsat: (lnwire.MilliSatoshi) 4000000000 mSAT,
     ExtraOpaqueData: ([]uint8) {
     }
    })
    
But i see other ChannelUpdate messages about fee of this channel to other nodes:
    
    
    2019-04-06 20:31:50.586 [TRC] PEER: writeMessage to 02a81c886f7cc867205c4fd4e6e9d4049a60db183599ab006d45e0c7170547a7ec@YY.YY.YY.YY:9735: (*lnwire.ChannelUpdate)(0xc05f2ba2c0)({
     Signature: (lnwire.Sig) (len=64 cap=64) {
      00000000  5d 52 dc 62 bf bc 8b 5e  bb 84 db cf a4 b6 89 0a  |]R.b...^........|
      00000010  04 60 5b 12 61 fb fb 3c  41 4a 53 b3 18 4b 9b fb  |.`[.a..<AJS..K..|
      00000020  0f c0 a2 77 7d 26 f7 74  9c 4b d3 75 4a 8c 1b 21  |...w}&.t.K.uJ..!|
      00000030  81 91 43 f3 3f 1b d2 5d  5b 09 0b f8 7e db d3 c3  |..C.?..][...~...|
     },
     ChainHash: (chainhash.Hash) (len=32 cap=32) 000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f,
     ShortChannelID: (lnwire.ShortChannelID) 570351:887:1,
     Timestamp: (uint32) 1554575479,
     MessageFlags: (lnwire.ChanUpdateMsgFlags) 00000001,
     ChannelFlags: (lnwire.ChanUpdateChanFlags) 00000001,
     TimeLockDelta: (uint16) 30,
     HtlcMinimumMsat: (lnwire.MilliSatoshi) 1000 mSAT,
     BaseFee: (uint32) 0,
     FeeRate: (uint32) 500,
     HtlcMaximumMsat: (lnwire.MilliSatoshi) 4000000000 mSAT,
     ExtraOpaqueData: ([]uint8) {
     }
    })
    
    2019-04-06 20:32:32.259 [TRC] PEER: writeMessage to 03e61043f5a1ad1c8903007290af131eb0a2d755d9e99848b9217a87482867122e@ZZ.ZZ.ZZ.ZZ:9735: (*lnwire.ChannelUpdate)(0xc05f2ba2c0)({
     Signature: (lnwire.Sig) (len=64 cap=64) {
      00000000  5d 52 dc 62 bf bc 8b 5e  bb 84 db cf a4 b6 89 0a  |]R.b...^........|
      00000010  04 60 5b 12 61 fb fb 3c  41 4a 53 b3 18 4b 9b fb  |.`[.a..<AJS..K..|
      00000020  0f c0 a2 77 7d 26 f7 74  9c 4b d3 75 4a 8c 1b 21  |...w}&.t.K.uJ..!|
      00000030  81 91 43 f3 3f 1b d2 5d  5b 09 0b f8 7e db d3 c3  |..C.?..][...~...|
     },
     ChainHash: (chainhash.Hash) (len=32 cap=32) 000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f,
     ShortChannelID: (lnwire.ShortChannelID) 570351:887:1,
     Timestamp: (uint32) 1554575479,
     MessageFlags: (lnwire.ChanUpdateMsgFlags) 00000001,
     ChannelFlags: (lnwire.ChanUpdateChanFlags) 00000001,
     TimeLockDelta: (uint16) 30,
     HtlcMinimumMsat: (lnwire.MilliSatoshi) 1000 mSAT,
     BaseFee: (uint32) 0,
     FeeRate: (uint32) 500,
     HtlcMaximumMsat: (lnwire.MilliSatoshi) 4000000000 mSAT,
     ExtraOpaqueData: ([]uint8) {
     }
    })

### LINKED ISSUE #2907: Private ChannelUpdate is not sent after funding gets 4 confirmations
Users were complaining about not being able to receive so I've tried to open two channels to LND `0.5.2-99-beta commit=queue/v1.0.1-269-gfa73a60167aa07df948020eb5f1f31893a113f9c`, here are my findings:

For the first channel BLW was kept online, remote peer started sending a private `ChannelUpdate` on reconnection once funding got its first confirmation (is it not too soon by the way?) and stopped doing that once funding had 4 confirmations.

For the second channel I closed BLW once funding was published and opened it when it had 10 confirmations, no `ChannelUpdate` was sent at all. If this is how it is supposed to work it's hardly a good fix for https://github.com/lightningnetwork/lnd/issues/1347 as mobile users often don't keep their devices online for a long time.

### PR DESCRIPTION
**Pull request by Roasbeef**
In this PR, we add a new test case to exercise a recent bug fix to
ensure that we no longer broadcast private channel policy changes. Along
the way, a few helper functions were added to slim down the test to the
core logic compared to some of the existing tests in this package. In
the future, these new helper functions should be utilized more widely for
tests in this package in order to cut down on some of the duplicated
logic.

Fixes #2904. 
Fixes #2907.

### CODE REVIEW

File: discovery/gossiper_test.go [hunk 1]
```diff
@@ -202,6 +202,27 @@ func (r *mockGraphSource) ForEachNode(func(node *channeldb.LightningNode) error)
...
+		info := info
+
+		edgeInfo := chans[info.ChannelID]
+		edgeInfo.Info = &info
+		chans[info.ChannelID] = edgeInfo
+	}
+	for _, edges := range r.edges {
+		edges := edges
+
+		edge := chans[edges[0].ChannelID]
```
Comment (cfromknecht) [inline discovery/gossiper_test.go]: is the source always the smaller of the two pubkeys?

File: discovery/gossiper_test.go [hunk 2]
```diff
@@ -3235,6 +3263,161 @@ func TestSendChannelUpdateReliably(t *testing.T) {
...
+func assertBroadcastMsg(t *testing.T, ctx *testCtx,
+	predicate func(lnwire.Message) error) {
+
+	t.Helper()
+
+	// We don't care about the order of the broadcast, only that our target
+	// predicate returns true for any of the messages, so we'll continue to
+	// retry until either we hit our timeout, or it returns with no error
+	// (message found).
+	err := lntest.WaitNoError(func() error {
```
Comment (cfromknecht) [inline discovery/gossiper_test.go]: 👍

File: discovery/gossiper_test.go [hunk 3]
```diff
@@ -202,6 +202,27 @@ func (r *mockGraphSource) ForEachNode(func(node *channeldb.LightningNode) error)
 
 func (r *mockGraphSource) ForAllOutgoingChannels(cb func(i *channeldb.ChannelEdgeInfo,
 	c *channeldb.ChannelEdgePolicy) error) error {
+
+	chans := make(map[uint64]channeldb.ChannelEdge)
+	for _, info := range r.infos {
```
Comment (wpaulino) [inline discovery/gossiper_test.go]: Missing mutex lock.

File: discovery/gossiper_test.go [hunk 4]
```diff
@@ -202,6 +202,27 @@ func (r *mockGraphSource) ForEachNode(func(node *channeldb.LightningNode) error)
...
+		info := info
```
Comment (wpaulino) [inline discovery/gossiper_test.go]: Nit: why is this necessary?

File: discovery/gossiper_test.go [hunk 5]
```diff
@@ -3263,6 +3263,161 @@ func TestSendChannelUpdateReliably(t *testing.T) {
...
+
+	// We don't care about the order of the broadcast, only that our target
+	// predicate returns true for any of the messages, so we'll continue to
+	// retry until either we hit our timeout, or it returns with no error
+	// (message found).
+	err := lntest.WaitNoError(func() error {
+		select {
+		case msg := <-ctx.broadcastedMessage:
+			return predicate(msg.msg)
+		case <-time.After(time.Second):
```
Comment (wpaulino) [inline discovery/gossiper_test.go]: Looks like this case will never be hit since the predicate timeout is shorter.

File: discovery/gossiper_test.go (hunk 4 above)
Comment (Roasbeef) [reply discovery/gossiper_test.go]: If we don't re-bind to the loop variable, then the pointer below will end up pointing to the same channel within the map.

Comment (wpaulino) [reply discovery/gossiper_test.go]: There's no pointer 😛

File: discovery/gossiper_test.go (hunk 1 above)
Comment (Roasbeef) [reply discovery/gossiper_test.go]: Hmm, the way we have things set up atm, we always just assume that the gossiper itself is the "first" key. I think this would only be an issue if we had an actual router sub-system hooked up, as that's what would commit this ordering to the index. FWIW, all the existing tests seem to be built around this assumption, and they all seem to run relatively reliably. I think this is in part due to the deterministic way we always assign ourselves to the first slot, and the remote party to the second slot.

**Comment by Roasbeef** [comment]
PTAL @cfromknecht @wpaulino, fix for #2904 is also in.

File: discovery/gossiper_test.go (hunk 1 above)
Comment (cfromknecht) [reply discovery/gossiper_test.go]: sounds good, the keys are generated randomly so we should know pretty quickly if this assumption is ever invalidated

File: discovery/gossiper.go [hunk 6]
```diff
@@ -1292,6 +1292,26 @@ func (d *AuthenticatedGossiper) processChanPolicyUpdate(
...
+		// send the update directly to the remote party.
+		if edgeInfo.info.AuthProof == nil {
+			remotePubKey := remotePubFromChanInfo(
+				edgeInfo.info, chanUpdate.ChannelFlags,
+			)
+			err := d.reliableSender.sendMessage(
+				chanUpdate, remotePubKey,
+			)
+			if err != nil {
+				log.Errorf("unable to reliably send %v for "+
```
Comment (cfromknecht) [inline discovery/gossiper.go]: nit: log message should be capitalized

File: discovery/utils.go [hunk 7]
```diff
@@ -146,3 +146,19 @@ func SignAnnouncement(signer lnwallet.MessageSigner, pubKey *btcec.PublicKey,
 
 	return signer.SignMessage(pubKey, data)
 }
+
+// remotePubFromChanInfo returns the public key of the remote peer given a
+// ChannelEdgeInfo that describe a channel we have with them.
+func remotePubFromChanInfo(chanInfo *channeldb.ChannelEdgeInfo,
+	msgFlags lnwire.ChanUpdateChanFlags) [33]byte {
```
Comment (cfromknecht) [inline discovery/utils.go]: nit: argument should be `chanFlags` as `msgFlags` would indicate a different field

File: discovery/gossiper.go (hunk 6 above)
Comment (Roasbeef) [reply discovery/gossiper.go]: What log message? We never capitalize log messages typically.

Comment (cfromknecht) [reply discovery/gossiper.go]: we've been applying this heuristic for the past 8 months at least

File: discovery/gossiper_test.go [hunk 8]
```diff
@@ -3266,6 +3266,161 @@ func TestSendChannelUpdateReliably(t *testing.T) {
...
+			}
+			if upd.TimeLockDelta != newTimeLockDelta {
+				return fmt.Errorf("wrong delta: expected %v, "+
+					"got %v", newTimeLockDelta,
+					upd.TimeLockDelta)
+			}
+
+			return nil
+		})
+	}
```
Comment (halseth) [inline discovery/gossiper_test.go]: should check that no more messages are broadcast here, as the private update could still be broadcast without the above loop catching it.
